### ISSUE TITLE
DT memory consumption grows over time and eventually CPU suddenly goes permanently to 100%

### ISSUE THREAD

**Issue by flemminglau** [DependencyTrack/dependency-track#2785]
### Current Behavior

We are seeing after some weeks that the memory consumption of DT gradually grows beyond 10Gb.
We also just had to restart DT as the APIServer had stopped responding in a sensible time and consuming 100% CPU.

We were also seeing timeout communicating w. the postgres server but probably this was a side effect of the lack of available CPU cycles.

### Steps to Reproduce

1. Load 107 projects
2. Leave DT running for 3 weeks uploading BOM updates for existing projects a few times a day.
3. Run scripts once a day exporting BOMs and VEXs.

### Expected Behavior

Memory consumption remains limited/constant

CPU consumption stays reasonable and never jumps to 100% permanently

### Dependency-Track Version

4.8.0

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by nscuro** [comment]
Are you [monitoring](https://docs.dependencytrack.org/getting-started/monitoring/) your instance? Any metrics that jump out in particular?

Also, how much memory did you allocate for the API server?

**Comment by flemminglau** [comment]
Unfortunately no.

We have not seen the need till now.

```
  deploy:
     resources:
       limits:
         memory: 12288m
       reservations:
         memory: 8192m
```
       
so allocation is 12Gb max.

We are not attempting to limit the JVM memory consumption.

**Comment by nscuro** [comment]
Also, what's the resource allocation for your database server? And did you configure database connection pooling to non-default values? https://docs.dependencytrack.org/getting-started/database-support/#connection-pooling

**Comment by flemminglau** [comment]
It is not so much the fact that a lot of ram is consumed that worries us.
More the incident of suddenly consuming all available CPU resources.

I have indeed due to some past DB issue updated the DB pool setting in my compose file to:

```
      ALPINE_DATABASE_POOL_MAX_SIZE: 40
      ALPINE_DATABASE_POOL_MIN_IDLE: 10
``` 

In the logs I see a few error messages logged. Sometimes 2-3 of these:
```
2023-05-25 01:58:22,702 ERROR [GlobalExceptionHandler] Uncaught internal server error
javax.ws.rs.NotFoundException: HTTP 404 Not Found
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
...
```

One complaining that I am trying to store a too long PURL

Some errors like this:
```
2023-04-28 17:30:20,934 ERROR [MavenMetaAnalyzer] Request failure
org.xml.sax.SAXParseException: Premature end of file.
        at org.apache.xerces.parsers.DOMParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.DocumentBuilderImpl.parse(Unknown Source)
        at java.xml/javax.xml.parsers.DocumentBuilder.parse(Unknown Source)
```
and like this
```
2023-04-25 17:40:04,595 ERROR [MavenMetaAnalyzer] Request failure
java.net.SocketTimeoutException: Read timed out
        at java.base/sun.nio.ch.NioSocketImpl.timedRead(Unknown Source)
        at java.base/sun.nio.ch.NioSocketImpl.implRead(Unknown Source)
        at java.base/sun.nio.ch.NioSocketImpl.read(Unknown Source)

```

Our network is a bit unstable sometimes which maybe explains some of these but still it should not be causing memory or CPU problems.

We will see if we can test the GC settings.

**Comment by flemminglau** [comment]
Managed to get monitoring working but unfortunately on a laptop VM which is not running 24x7.

Today we again (after 3.5 weeks of heap consumption gradually increasing to 8Gb) had to restart DT as it was looping consuming 350% CPU.

Looking at the monitoring I notice a few things that look abnormal:
![image](https://github.com/DependencyTrack/dependency-track/assets/25148089/d2ef822d-b929-4c8d-90ca-7259f6697ffe)

I see "Database operations", "Events published" and "Notifications Published" skyrocketing yesterday afternoon.    

I can dig further into this historic data if you can tell me what to look for.

**Comment by nscuro** [comment]
Did the spike happen when the system was supposedly running idle?

It looks like it might be related to rebuilding of the search indexes. Can you check the logs for that time frame, and specifically look for messages from `IndexManager`? If indexes are being rebuilt, you'll see something like `Starting reindex task` being logged.

**Comment by flemminglau** [comment]
I think you are right.
The spike I see is indeed at the time when the log has:
```
2023-06-19 17:33:05,077 INFO [IndexManager] Delta ratio for type VULNERABLESOFTWARE : 2.0%
2023-06-19 17:33:05,077 INFO [ComponentIndexer] Starting reindex task. This may take some time.
2023-06-19 17:33:05,077 INFO [IndexManager] Deleting component index
2023-06-19 17:33:05,119 INFO [ComponentIndexer] Indexing 829 components in project: 66d0a4f0-20ba-4a3b-b136-103b63a29589
2023-06-19 17:33:05,123 INFO [ComponentIndexer] Completed indexing of 829 components in project: 66d0a4f0-20ba-4a3b-b136-103b63a29589
```



It is actually (looking in the log) not until 03:10:59 that things start looking ugly in the log.
There I see:
```
2023-06-22 03:09:06,126 INFO [OssIndexAnalysisTask] Sonatype OSS Index analysis complete
2023-06-22 03:09:10,956 INFO [PolicyEngine] Evaluating 1 component(s) against applicable policies
2023-06-22 03:10:59,466 WARN [HikariPool] non-transactional - Thread starvation or clock leap detected (housekeeper delta=53s615ms856ns).
2023-06-22 03:11:10,377 INFO [RepositoryMetaAnalyzerTask] Performing component repository metadata analysis against 1 components
2023-06-22 03:13:00,780 WARN [HikariPool] non-transactional - Thread starvation or clock leap detected (housekeeper delta=54s508ms159µs119ns).
2023-06-22 03:12:48,885 ERROR [GlobalExceptionHandler] Uncaught internal server error
java.lang.IllegalArgumentException: HV000116: The object to be validated must not be null.
        at org.hibernate.validator.internal.util.Contracts.assertNotNull(Contracts.java:44)
        at org.hibernate.validator.internal.engine.ValidatorImpl.validateProperty(ValidatorImpl.java:177)
        at org.dependencytrack.resources.v1.ComponentResource.updateComponent(ComponentResource.java:335)
        at jdk.internal.reflect.GeneratedMethodAccessor616.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:134)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:177)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:81)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
```

From then on this same error keeps re-appearing.

It may of course be a side effect of CPU resources being exhausted by something else which is not being logged.

Unfortunately I do not have monitoring data until much later.

**Comment by Shawyeok** [comment]
I've encounter similar problem, inspect apiserver process with `top -H -p <pid>` command, it shows that the GC threads took all cpu. I've made a heapdump which is quite big, after analyze it I found a `org.datanucleus.api.jdo.JDOPersistenceManagerFactory` object took almost 10GB memory, here are some screenshots of `mat`:
![image](https://github.com/DependencyTrack/dependency-track/assets/5058708/30181877-d297-455b-ac30-bd792e6f4b60)
![image](https://github.com/DependencyTrack/dependency-track/assets/5058708/a4f2dc86-0bc7-44f1-9c5f-9439a88c5e3f)

So, basically I'm pretty sure there is a memory leak issue in apiserver (`4.8.0` which is I'm using), now restart the apiserver is a workaound solution, I'd like to open a PR later if it possible.

@flemminglau @nscuro

**Comment by nscuro** [comment]
@Shawyeok That is indeed interesting. `pmCache` according to the DataNucleus source code holds references to all active `PersistenceManager`s (in DT this correlates with active `QueryManager`s). They should be rather short-lived, and their respective entries in `pmCache` should never stay around for long.

Are you able to correlate the increase of memory usage with some tasks being executed, an increase in activity on the system, or is it just continuously growing over time?

**Comment by Shawyeok** [comment]
In `org.dependencytrack.policy.AbstractPolicyEvaluator`, there is a `QueryManager` created while instance init, https://github.com/DependencyTrack/dependency-track/blob/a7e512235c33b650decc90a8c12ceedc5234c6d9/src/main/java/org/dependencytrack/policy/AbstractPolicyEvaluator.java#L37-L41

and later overwrite by it's setter which is called in `VulnerabilityAnalysisTask` execution, I believe it's the cause I've encounter, WDYT?
https://github.com/DependencyTrack/dependency-track/blob/a7e512235c33b650decc90a8c12ceedc5234c6d9/src/main/java/org/dependencytrack/policy/PolicyEngine.java#L88

**Comment by nscuro** [comment]
@Shawyeok Woah, amazing spot! Yes, this indeed looks like it could cause the issue. Would you like to raise a PR with a fix, to claim the credit for your discovery?

**Comment by Shawyeok** [comment]
> @Shawyeok Woah, amazing spot! Yes, this indeed looks like it could cause the issue. Would you like to raise a PR with a fix, to claim the credit for your discovery?

Thanks for asking, I'll do it later.

**Comment by Shawyeok** [comment]
@nscuro 
I've identified another location that could potentially cause a memory leak issue. I've created two separate PRs to address this. Please take a look.

#2872 #2873

**Comment by nscuro** [comment]
@Shawyeok Absolutely correct, that too looks like a proper resource leak. Thanks for finding and fixing!

**Comment by nscuro** [comment]
Closing as we are confident that the implemented improvements indeed fix this issue. If the problem persists after upgrading to 4.9, memory leaks in other areas of the code would be responsible, and those we'll deal with in separate issues.
